src/main/java/com/adobe/epubcheck/ctc/EpubSVGCheck.java

```diff
@@ -123,9 +123,22 @@ void checkViewBox(String svgDocEntry, Document doc)
     {
       Element svgElement = (Element) n.item(i);
       String viewport = svgElement.getAttributeNS(svgNS, "viewBox");
-      if (viewport == null || viewport.length() == 0)
+      boolean viewportFound = (viewport != null && viewport.length() > 0);
+      String height = svgElement.getAttributeNS(svgNS,"height");
+      boolean heightFound = (height != null && height.length() > 0);
+      String width = svgElement.getAttributeNS(svgNS,"width");
+      boolean widthFound = (width != null && width.length() > 0);
+      //if (viewport == null || viewport.length() == 0)
+      if (!viewportFound)
       {
-        report.message(MessageId.HTM_048, EPUBLocation.create(svgDocEntry, XmlDocParser.getElementLineNumber(svgElement), XmlDocParser.getElementColumnNumber(svgElement)));
+      	if (!heightFound || !widthFound){
+          report.message(MessageId.HTM_048, EPUBLocation.create(svgDocEntry, XmlDocParser.getElementLineNumber(svgElement), XmlDocParser.getElementColumnNumber(svgElement)));
+		}else{
+          report.message(MessageId.HTM_054, EPUBLocation.create(svgDocEntry, XmlDocParser.getElementLineNumber(svgElement), XmlDocParser.getElementColumnNumber(svgElement)),svgDocEntry);
+	      if (width.contains("%") || height.contains("%")){
```

Comment:
I think this test is too permissive: the spec says the values should be defined in **pixel** values, i.e. all other non-percent units should trigger the warning too.

I suggest we change the test by matching to a px-value regex (something like `\d+px|0`)

Reply:
Right, I think you could just make the test \d+(px)? as pixels are the default unit, or at least that's what I get out of this para:

> For the outermost svg element, the SVG user agent must determine an initial viewport coordinate system and an initial user coordinate system such that the two coordinates systems are identical. The origin of both coordinate systems must be at the origin of the SVG viewport, and one unit in the initial coordinate system must equal one CSS 2.1 px ([CSS2], section 4.3.2) in the SVG viewport.

Reply:
Yeah I'm not sure: CSS values says `For zero lengths the unit identifier is optional`; I don't think it's allowed to have unit-less length values other than "0".

Reply:
In the units section of SVG 1.1 it says:

> All coordinates and lengths in SVG can be specified with or without a unit identifier.
>
> When a coordinate or length value is a number without a unit identifier (e.g., "25"), then the given coordinate or length is assumed to be in user units (i.e., a value in the current user coordinate system).

I pulled that other quote out of SVG2, I think, as I followed the EPUB reference, but I don't think it's changed that the default user unit is pixel in these cases.

Reply:
Right, I was just looking at SVG 2 and it says:

> If the width or height presentation attributes on the outermost svg element are in user units (i.e., no unit identifier has been provided), then the value is assumed to be equivalent to the same number of "px" units (see Units).

So let's go for `\d+(px)?`!